src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/NumericIndexWriter.java

```diff
@@ -0,0 +1,179 @@
+ * buffer all point values to sort them.
+ */
+public class NumericIndexWriter
+{
+    public static final int MAX_POINTS_IN_LEAF_NODE = BlockBalancedTreeWriter.DEFAULT_MAX_POINTS_IN_LEAF_NODE;
+
+    private final BlockBalancedTreeWriter writer;
+    private final IndexDescriptor indexDescriptor;
+    private final IndexContext indexContext;
+    private final int bytesPerValue;
+
+    /**
+     * @param maxSegmentRowId maximum possible segment row ID, used to create `maxDoc` for kd-tree
+     */
+    public NumericIndexWriter(IndexDescriptor indexDescriptor,
+                              IndexContext indexContext,
+                              int bytesPerValue,
+                              long maxSegmentRowId)
+    {
+        this(indexDescriptor, indexContext, MAX_POINTS_IN_LEAF_NODE, bytesPerValue, maxSegmentRowId);
+    }
+
+    @VisibleForTesting
+    public NumericIndexWriter(IndexDescriptor indexDescriptor,
+                              IndexContext indexContext,
+                              int maxPointsInLeafNode,
+                              int bytesPerValue,
+                              long maxSegmentRowId)
+    {
+        checkArgument(maxSegmentRowId >= 0, "[%s] maxRowId must be non-negative value, but got %s", indexContext.getIndexName(), maxSegmentRowId);
+
+        this.indexDescriptor = indexDescriptor;
+        this.indexContext = indexContext;
+        this.bytesPerValue = bytesPerValue;
+        this.writer = new BlockBalancedTreeWriter(maxSegmentRowId + 1, bytesPerValue, maxPointsInLeafNode);
+    }
+
+    @Override
+    public String toString()
+    {
+        return MoreObjects.toStringHelper(this)
+                          .add("bytesPerDim", bytesPerValue)
+                          .add("bufferedPoints", writer.getPointCount())
+                          .toString();
+    }
+
+    public static class LeafCallback implements BlockBalancedTreeWriter.OneDimensionBKDWriterCallback
+    {
+        final List<PackedLongValues> postings = new ArrayList<>();
+
+        public int numLeaves()
+        {
+            return postings.size();
+        }
+
+        @Override
+        public void writeLeafDocs(BlockBalancedTreeWriter.RowIDAndIndex[] sortedByRowID, int offset, int count)
+        {
+            PackedLongValues.Builder builder = PackedLongValues.monotonicBuilder(PackedInts.COMPACT);
+
+            for (int i = offset; i < count; ++i)
+            {
+                builder.add(sortedByRowID[i].rowID);
+            }
+            postings.add(builder.build());
+        }
+    }
+
+    /**
+     * Writes a k-d tree and posting lists from a {@link org.apache.lucene.codecs.MutablePointValues}.
+     *
+     * @param values points to write
+     *
+     * @return metadata describing the location and size of this kd-tree in the overall SSTable kd-tree component file
+     */
+    public SegmentMetadata.ComponentMetadataMap writeAll(IntersectingPointValues values) throws IOException
```

Comment:
This method has multiple vars using the `bkd` prefix. Maybe we could just refer to the block balanced tree in the context of this class as just "the tree", and name those vars `treePosition`, `treeOutput`, `treeOffset`, `treeLength`, etc.

Reply:
I've gone with `balancedTree` as a prefix for these. Tbh, I'm moving towards using `BalancedTree` and the class prefix and `balancedtree` as the package now. 

I'm going to continue with the renaming process as is and we can look at it again when it is complete.

Reply:
`balancedTree` as prefix for tree things sounds good to me. One less word, and repeating the `block` part once and again in a tree structure inside the disk package doesn't add too much value.